src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1726,4 +1726,22 @@ impl Cage {
             syscall_error(Errno::EBADF, "getdents", "Invalid file descriptor")
         }
     }
+
+    //------------------------------------GETCWD SYSCALL------------------------------------
+    
+    pub fn getcwd_syscall(&self, buf: *mut u8, bufsize: u32) -> i32 {
+        let cwd: String = format!("{}{}", self.cwd.read().unwrap().to_string_lossy(), "\0");
+        let bytes: Vec<u8> = cwd.as_bytes().to_vec();
+        let length = bytes.len();
+
+        use std::convert::TryInto;
+        if bufsize < length.try_into().unwrap() {
+            return syscall_error(Errno::ERANGE, "getcwd", "the length (in bytes) of the absolute pathname of the current working directory exceeds the given size");
+        }
+        
+        let slice = unsafe{std::slice::from_raw_parts_mut(buf, length)};
```

Comment:
We additionally don't use any unsafe code except for in the interface, so this should probably be handled into the dispatcher. I'd say that this function should probably take in an `&mut [u8]`, which you can get from `get_slice_from_string`, which you should use in the dispatcher. Check the implementation of Epoll for something similar we did, although the slice was not of u8s, it should be relatively easy to understand how to transform these things.

Reply:
The epoll implementations are in the sockremoval_better branch

Reply:
Updated this with the last few commits.